### PR TITLE
Callback for tracing sql queries

### PR DESCRIPTION
**Pull request by smacker**
To solve #1281 I came up with this solution...

### CODE REVIEW

**Comment by jinzhu** [comment]
Hi @smacker 

Thank you for your PR, sorry just noticed your issue report, I think you could just make it works with set a new logger, so this might be not necessary.

Refer: https://github.com/jinzhu/gorm/blob/master/logger.go

Thank you.

**Comment by smacker** [comment]
No, it won't work with a logger. I have to start newrelic transaction before sql started execution and finish it after. It measures not only time, but call stack, how much cpu it burn and other things.

**Comment by smacker** [comment]
please reconsider trace callback. Code instrumentation is very useful. If you don't like callback, can you suggest any other way to do it? For example go-redis provides WrapProcess method: https://github.com/go-redis/redis/pull/352

File: scope.go [hunk 1]
```diff
@@ -947,6 +947,9 @@ func (scope *Scope) typeName() string {
 
 // trace print sql log
 func (scope *Scope) trace(t time.Time) {
+	scope.Set("gorm:trace-time", t)
```
Comment (zardak) [inline scope.go]: IMO, this is a sign of hacky approach. The more or less right way would be to introduce lifecycle hooks, like `BeforeX` and `AfterX` callbacks.

Comment (smacker) [inline scope.go]: agreed. I'll try to improve it.

**Comment by ipadavic** [comment]
I'm trying to add sql tracing to my app via datadog apm, this PR makes it possible. Any chance this could be merged?

**Comment by zardak** [comment]
@jinzhu could you take a look at the pr please?

**Comment by jinzhu** [comment]
Hi @smacker 

Sorry for the delay, just have time to review it and your code in https://github.com/smacker/newrelic-context/blob/master/gorm.go

I think for your problem you could resolve it with `RawQuery` callback, and it will be call in `Row`, `Rows`, `Count`...   Refer https://github.com/jinzhu/gorm/blob/master/callback.go#L78

**Comment by smacker** [comment]
@jinzhu it was my first intention to use raw callback, but you call it before sql.DB methods, there is no way to know when it finish. For Query, Create, Delete and Update you call sql.DB inside callbacks, so I can add my callback before and after real query. But for RawQuery you do sql query inside scope, not callback. I also tried to change Row, Rows, Count, ... methods and move sql related code from scope to inside callbacks, but code becomes quite messy.

**Comment by jinzhu** [comment]
Hi @smacker 

Thank you for your report, just fixed that problem. https://github.com/jinzhu/gorm/commit/c62e9bcabea2b3ed5e3a8bc1602cd38f8ad477f9

**Comment by smacker** [comment]
@jinzhu your commit broke the code of anybody who altered the query in row callbacks. (your new callback registered before any existed one) JFK

**Comment by jinzhu** [comment]
Yes, you are right, just pushed another commit to fix the behaviour with a warning message https://github.com/jinzhu/gorm/commit/7fb9b62c17d90320e0582a4720db025c1652fd6a

Thank you for the report.

**Comment by ipadavic** [comment]
Sorry for commenting on closed PR, but I'm having trouble getting this to work with Raw(). Does "row_query" callback include Raw queries?

**Comment by jinzhu** [comment]
No, `Raw` query is really raw, no callback included.

Only `Row`, `Rows`, `Count`, `Plack`... those methods include the callback.

**Pull request closed without merging**